// zcatalyst-cli/lib/serve/server/lib/node/server/bio.js
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { IncomingMessage, Server, ServerResponse } from 'http';
import { inspect } from 'util';

const endTime = 30 * 1000 + Date.now();

/**
 *
 * @param {IncomingMessage} req
 * @param {ServerResponse} res
 * @returns
 */
function buildContext(req, res, locals) {
	const catalystHeaders = {};
	Object.entries(req.headers).forEach(([header, value]) => {
		const _header = header.toLowerCase();
		if (_header.startsWith('x-zc-')) {
			catalystHeaders[_header] = value;
		}
	});
	return {
		catalystHeaders,
		getMaxExecutionTimeMs: () => 30 * 1000, // 30 sec
		getRemainingExecutionTimeMs: () => endTime - Date.now(),
		close: () => {
			const _finalRes = JSON.stringify({
				output: locals.finalResponse || ''
			});
			res.setHeader('Content-Type', 'application/json');
			res.setHeader('Content-Length', _finalRes.length);
			res.write(_finalRes);
		},
		// eslint-disable-next-line no-console
		log: (...params) => console.log(...params)
	};
}

/**
 *
 * @param {IncomingMessage} req
 * @param {ServerResponse} res
 * @returns
 */
async function buildBasicIO(req, res, locals) {
	const urlParts = new URL(`http://${req.headers.host}${req.url}`).searchParams;
	const userData = JSON.stringify(Object.fromEntries(urlParts.entries()));

	const body = await new Promise((resolve, reject) => {
		const buff = [];
		req.on('data', (ch) => buff.push(ch));
		req.on('end', () => {
			try {
				const _jsonStr = Buffer.concat(buff).toString();
				resolve(_jsonStr ? JSON.parse(_jsonStr) : {});
			} catch (er) {
				reject(er);
			}
		});
		req.on('error', reject);
	});
	const basicIO = {
		write: (VALUE) => {
			if (!typeof VALUE === 'string') {
				// eslint-disable-next-line no-console
				console.error(`Expected string but ${typeof VALUE} found!.`);
			} else {
				const _res = VALUE.substring(0, 1 * 1024);
				// eslint-disable-next-line no-console
				console.log(_res);
				locals.finalResponse += _res;
			}
		},
		getArgument: (KEY) => userData[KEY] || body[KEY],
		getAllArguments: () => (userData && Object.keys(userData).length > 0 ? userData : body),
		/**
		 * @deprecated use get getArgument instead.
		 */
		getParameter: (KEY) => {
			// eslint-disable-next-line no-console
			console.error(
				'CatalystDeprecationWarning: basicIO.getParameter() is deprecated due to usability issues. Please use basicIO.getArgument() method instead.'
			);
			return basicIO.getArgument(KEY);
		},
		/**
		 * @deprecated use get getAllArgument instead.
		 */
		getAllParameters: () => {
			// eslint-disable-next-line no-console
			console.error(
				'CatalystDeprecationWarning: basicIO.getAllParameters() is deprecated due to usability issues. Please use basicIO.getAllArgument() method instead.'
			);
			return basicIO.getAllArguments();
		},
		setStatus: (status) => {
			res.statusCode = status;
		}
	};
	return basicIO;
}

/**
 * BasicIO handler
 */
export default (index, targetName, type) => async (req, res) => {
	try {
		if (index === undefined) {
			throw new Error(`index of ${type} function is undefined`);
		}
		const userModule = await import(index);
		if (type !== 'bio') {
			throw new Error('Invalid URL pattern for the functions type: ' + type);
		}
		if (!('default' in userModule)) {
			throw new Error('Could not find any default export');
		}
		if (typeof userModule.default !== 'function') {
			throw new Error('The default export is not a function');
		}
		const locals = {
			finalResponse: ''
		};
		const context = buildContext(req, res, locals);
		const basicIO = await buildBasicIO(req, res, locals);
		await userModule.default(context, basicIO);
	} catch (e) {
		const errorString = inspect(e);
		// eslint-disable-next-line no-console
		console.error('[' + targetName + '] ' + errorString);
		res.writeHead(500);
		res.end(errorString);
	}
};
