// zcatalyst-cli/lib/serve/server/lib/slate/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const error_1 = __importDefault(require("../../../../error"));
const logger_1 = require("../../../../util_modules/logger");
const shell_1 = require("../../../../util_modules/shell");
const master_1 = __importDefault(require("../master"));
const file_names_1 = __importDefault(require("../../../../util_modules/constants/lib/file-names"));
const static_server_1 = __importDefault(require("./static-server"));
const open_1 = __importDefault(require("open"));
const async_1 = require("../../../../util_modules/fs/lib/async");
const startSlate = (port, opts) => {
    const child = (0, shell_1.spawn)(opts.command, [], {
        cwd: opts.target,
        shell: true,
        env: {
            ZC_SLATE_PORT: port + ''
        }
    }).RAW();
    return new Promise((_res, _rej) => {
        child.on('spawn', () => _res(child));
        child.on('error', (reason) => _rej(reason));
    });
};
exports.default = (serverDetails) => __awaiter(void 0, void 0, void 0, function* () {
    var _a;
    const targetSlate = serverDetails.target;
    if (!targetSlate || !targetSlate.config) {
        throw new error_1.default('Slate details not found');
    }
    let child;
    if (targetSlate.config.framework === 'static') {
        child = yield (0, static_server_1.default)(targetSlate.port.slate, targetSlate.source, {
            enableWatch: true
        });
        child.addListener('start', () => {
            (0, open_1.default)('http://localhost:' + targetSlate.port.slate).catch();
        });
    }
    else {
        const configFile = (0, path_1.join)(targetSlate.source, file_names_1.default.cli_config);
        const configJson = yield (0, async_1.readJSONFile)(configFile);
        const installDeps = (0, shell_1.spawn)('npm', ['install'], {
            cwd: targetSlate.source
        }).RAW();
        yield new Promise((resolve, reject) => {
            installDeps.on('exit', (code) => {
                var _a;
                if (code === 0) {
                    resolve();
                }
                else {
                    reject(new error_1.default('Failed to install dependencies', {
                        exit: 1,
                        errorId: 'SERVE-SLATE-3',
                        arg: [((_a = targetSlate.config) === null || _a === void 0 ? void 0 : _a.install_command) + ''],
                    }));
                }
            });
            installDeps.on('error', (err) => {
                var _a;
                reject(new error_1.default('Failed to install dependencies', {
                    exit: 1,
                    errorId: 'SERVE-SLATE-3',
                    arg: [((_a = targetSlate.config) === null || _a === void 0 ? void 0 : _a.install_command) + '', 'catalyst serve'],
                    original: err
                }));
            });
        });
        child = yield startSlate(targetSlate.port.slate, {
            target: targetSlate.source,
            command: (_a = configJson === null || configJson === void 0 ? void 0 : configJson.slate) === null || _a === void 0 ? void 0 : _a.dev_command
        });
    }
    const masterServe = yield (0, master_1.default)(targetSlate.port.proxy, {
        slateDetails: serverDetails
    });
    child.once('exit', (code) => {
        masterServe.close((err) => {
            if (err) {
                (0, logger_1.debug)('Error stopping the Slate proxy server: ', err);
            }
        });
        if (code === 150) {
            targetSlate.validity = {
                valid: false,
                reason: 'Unable to start the Slate'
            };
        }
    });
    process.on('SIGINT', () => {
        masterServe.close((err) => {
            if (err) {
                (0, logger_1.debug)('Error stopping the Slate proxy server: ', err);
            }
        });
        targetSlate.trigger.kill = true;
    });
    return child;
});
