// @zcatalyst/cli-utils/out/throbber/utils.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHAR = void 0;
exports.breakText = breakText;
exports.getLinesLength = getLinesLength;
exports.writeStream = writeStream;
exports.cleanStream = cleanStream;
exports.supportsUnicode = supportsUnicode;
const readline_1 = require("readline");
const strip_ansi_1 = __importDefault(require("strip-ansi"));
const env_1 = require("../env");
const ansi_colors_1 = require("ansi-colors");
function breakText(text, prefixLength) {
    return (0, strip_ansi_1.default)(text)
        .split('\n')
        .map((line, index) => (index === 0 ? breakLine(line, prefixLength) : breakLine(line, 0)))
        .join('\n');
}
function breakLine(line, prefixLength) {
    const columns = process.stderr.columns || 95;
    return line.length >= columns - prefixLength
        ? `${line.substring(0, columns - prefixLength - 1)}\n${breakLine(line.substring(columns - prefixLength - 1, line.length), 0)}`
        : line;
}
function getLinesLength(text, prefixLength) {
    return (0, strip_ansi_1.default)(text)
        .split('\n')
        .map((line, index) => (index === 0 ? line.length + prefixLength : line.length));
}
function writeStream(stream, output, rawLines) {
    stream.write(output);
    (0, readline_1.moveCursor)(stream, 0, -rawLines.length);
}
function cleanStream(stream, rawLines) {
    rawLines.forEach((lineLength, index) => {
        (0, readline_1.moveCursor)(stream, lineLength, index);
        (0, readline_1.clearLine)(stream, 1);
        (0, readline_1.moveCursor)(stream, -lineLength, -index);
    });
    (0, readline_1.moveCursor)(stream, 0, rawLines.length);
    (0, readline_1.clearScreenDown)(stream);
    (0, readline_1.moveCursor)(stream, 0, -rawLines.length);
}
// fn's ported from other modules
const isSupported = !env_1.isWindows || process.env.CI || process.env.TERM === 'xterm-256color';
exports.CHAR = {
    success: isSupported ? ansi_colors_1.bold.green('✔') : ansi_colors_1.bold.green('√'),
    warning: isSupported ? ansi_colors_1.bold.yellow('⚠') : ansi_colors_1.bold.yellow('‼'),
    info: isSupported ? ansi_colors_1.bold.cyan('ℹ') : ansi_colors_1.bold.cyan('i'),
    error: isSupported ? ansi_colors_1.bold.red('✖') : ansi_colors_1.bold.red('×')
};
function supportsUnicode() {
    // The default command prompt and powershell in Windows do not support Unicode characters.
    // However, the VSCode integrated terminal and the Windows Terminal both do.
    return (process.platform !== 'win32' ||
        process.env.TERM_PROGRAM === 'vscode' ||
        !!process.env.WT_SESSION);
}
