// zcatalyst-cli/lib/serve/server/lib/appsail/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const error_1 = __importDefault(require("../../../../error"));
const constants_1 = require("../../../../util_modules/constants");
const fs_1 = require("../../../../util_modules/fs");
const index_1 = require("../../../../util_modules/logger/index");
const project_1 = require("../../../../util_modules/project");
const shell_1 = require("../../../../util_modules/shell");
const endpoints_1 = require("../../../../endpoints");
const archiver_1 = __importDefault(require("../../../../archiver"));
const option_1 = require("../../../../util_modules/option");
const execute_script_1 = require("../../../../execute-script");
const master_1 = __importDefault(require("../master"));
const ansi_colors_1 = require("ansi-colors");
const common_1 = require("../../../../init/util/common");
const utils_1 = require("../master/utils");
const port_resolver_1 = __importDefault(require("../../../../port-resolver"));
const container_1 = require("../../../../util_modules/container");
const container_plugin_1 = require("@zcatalyst/container-plugin");
const appsail_1 = require("../../../../util_modules/config/lib/appsail");
const image_1 = require("@zcatalyst/container-plugin/out/endpoints/image");
function executeHook(script, name, moduleSource) {
    return __awaiter(this, void 0, void 0, function* () {
        if ((0, option_1.getOptionValue)('ignoreScripts', false)) {
            (0, index_1.debug)(`skipping ${name} hook`);
            return;
        }
        return (0, execute_script_1.executeCommand)(script, { moduleSource, feature: name });
    });
}
const startAppSail = (port, opts) => {
    if (port === -1) {
        throw new error_1.default('The AppSail Http port value is invalid: ' + port, { exit: 2 });
    }
    const startScriptPath = (0, path_1.join)(__dirname, 'start.js');
    const _opts = [startScriptPath, '-t', opts.type];
    if (opts.command) {
        _opts.push('-c', opts.command);
    }
    const child = (0, shell_1.spawn)('node', _opts, {
        cwd: opts.target,
        stdio: 'pipe',
        killSignal: 'SIGINT',
        env: Object.assign({ X_ZOHO_CATALYST_LISTEN_PORT: port + '', X_ZOHO_CATALYST_RUNTIME_MEMORY: opts.memory + '', X_ZOHO_CATALYST_ACCOUNTS_URL: constants_1.ORIGIN.auth, X_ZOHO_CATALYST_CONSOLE_URL: constants_1.ORIGIN.admin, X_ZOHO_STRATUS_RESOURCE_SUFFIX: constants_1.ORIGIN.stratusSuffix, X_ZOHO_CATALYST_IS_LOCAL: 'true', X_ZC_RESOURCE_NAME: opts.name + '', CATALYST_PORTAL_DOMAIN: constants_1.ORIGIN.iamPortal, CATALYST_PROJECT_TIMEZONE: (0, project_1.getProjectTimezone)(Intl.DateTimeFormat().resolvedOptions().timeZone) }, (opts.env || {}))
    }).RAW();
    return new Promise((_res, _rej) => {
        child.on('spawn', () => _res(child));
        child.on('error', (reason) => _rej(reason));
    });
};
function startAppSailWithContainer(serverDetails) {
    var _a, _b, _c, _d, _e, _f, _g;
    return __awaiter(this, void 0, void 0, function* () {
        const targetSail = serverDetails.target;
        if (!targetSail || !targetSail.config) {
            throw new error_1.default('Invalid AppSail target', { exit: 2 });
        }
        (0, index_1.debug)('Using container to serve the AppSail: ' + targetSail.name);
        targetSail.config.env_variables = Object.assign({ X_ZOHO_CATALYST_ACCOUNTS_URL: constants_1.ORIGIN.auth, X_ZOHO_CATALYST_CONSOLE_URL: constants_1.ORIGIN.admin, X_ZOHO_CATALYST_RESOURCE_ID: targetSail.name, X_ZOHO_STRATUS_RESOURCE_SUFFIX: constants_1.ORIGIN.stratusSuffix, X_ZOHO_CATALYST_CODE_LOCATION: '/catalyst', CATALYST_PROJECT_TIMEZONE: (0, project_1.getProjectTimezone)(Intl.DateTimeFormat().resolvedOptions().timeZone), X_ZOHO_CATALYST_ORG: (0, project_1.getEnvId)() || '', X_ZOHO_ADMIN_CRED_TOKEN: 'Unknown' }, (_a = targetSail === null || targetSail === void 0 ? void 0 : targetSail.config) === null || _a === void 0 ? void 0 : _a.env_variables);
        if (!((_b = serverDetails.target) === null || _b === void 0 ? void 0 : _b.source)) {
            throw new error_1.default('Invalid AppSail source: ' + targetSail.source, {
                exit: 1
            });
        }
        let child;
        switch (true) {
            case (_c = serverDetails.target) === null || _c === void 0 ? void 0 : _c.source.startsWith(appsail_1.CONTAINER_IMAGE_PROTOCOLS.dockerArchive): {
                const tarPath = serverDetails.target.source.replace(appsail_1.CONTAINER_IMAGE_PROTOCOLS.dockerArchive, '');
                const appsailTar = yield fs_1.ASYNC.fileExists(tarPath);
                if (!appsailTar) {
                    throw new error_1.default('Invalid AppSail source: ' + targetSail.source, {
                        exit: 1
                    });
                }
                const tarStream = fs_1.SYNC.getReadStream(tarPath);
                const imgName = `${serverDetails.target.name}-tar`;
                const imageInfo = yield (0, image_1.loadTempImageTar)(imgName, tarStream);
                serverDetails.target.source = 'docker://' + ((_d = imageInfo.RepoTags) === null || _d === void 0 ? void 0 : _d.at(0));
                child = yield (0, container_plugin_1.start)({
                    details: serverDetails,
                    projectId: (0, project_1.getProjectId)(),
                    ruok: container_plugin_1.RUOK.SKIP
                });
                child.once('exit', () => {
                    (0, image_1.removeImages)(imgName).catch((er) => (0, index_1.debug)('Error removing the temp images on exit: ', er));
                });
                child.once('error', (err) => {
                    (0, index_1.debug)('Error with the AppSail container: ', err);
                    (0, image_1.removeImages)(imgName).catch((er) => (0, index_1.debug)('Error removing the temp images on error: ', er));
                });
                break;
            }
            case serverDetails.target.runtime === 'catalyst': {
                const projectRoot = (0, project_1.getProjectRoot)();
                if (!((_f = (_e = serverDetails.target.config) === null || _e === void 0 ? void 0 : _e.build_path) === null || _f === void 0 ? void 0 : _f.startsWith(projectRoot))) {
                    throw new error_1.default('AppSail source outside project directory. For security reasons, Catalyst' +
                        'runtime AppSails served within containers should have the source withing the project root directory', {
                        exit: 1,
                        context: {
                            targetSource: serverDetails.target.source,
                            projectRoot
                        }
                    });
                }
            }
            case (_g = serverDetails.target) === null || _g === void 0 ? void 0 : _g.source.startsWith(appsail_1.CONTAINER_IMAGE_PROTOCOLS.docker): {
                child = yield (0, container_plugin_1.start)({
                    details: serverDetails,
                    projectId: (0, project_1.getProjectId)(),
                    ruok: container_plugin_1.RUOK.SKIP
                });
                break;
            }
            default: {
                throw new error_1.default('Invalid AppSail runtime/docker image', {
                    exit: 2,
                    context: serverDetails
                });
            }
        }
        child.emit('start', serverDetails);
        return child;
    });
}
exports.default = (serverDetails) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
    let child;
    const targetSail = serverDetails.target;
    if (!targetSail || !targetSail.config) {
        throw new error_1.default('AppSail details not found');
    }
    if ((targetSail === null || targetSail === void 0 ? void 0 : targetSail.runtime) === 'catalyst') {
        if ((_a = targetSail.config.scripts) === null || _a === void 0 ? void 0 : _a.preserve) {
            yield executeHook(targetSail.config.scripts.preserve, `AppSail [PRESERVE] [${targetSail.name}]`, targetSail.source);
        }
        const buildPath = (_b = targetSail.config) === null || _b === void 0 ? void 0 : _b.build_path;
        if (!buildPath) {
            throw new error_1.default('Invalid AppSail config: invalid build path :: ' + buildPath, {
                exit: 2
            });
        }
        const target = (yield fs_1.ASYNC.fileExists(buildPath)) ? (0, path_1.dirname)(buildPath) : buildPath;
        if (!(yield fs_1.ASYNC.isPathExists(target))) {
            if ((0, path_1.isAbsolute)(((_c = targetSail.config) === null || _c === void 0 ? void 0 : _c.raw.build_path) || '')) {
                throw new error_1.default('The given AppSail build path does not exists', {
                    exit: 1,
                    errorId: 'SERVE-APPSAIL-1',
                    arg: [ansi_colors_1.italic.underline.red(target), (0, ansi_colors_1.bold)(targetSail.name)]
                });
            }
            throw new error_1.default('The given AppSail build path does not exists: ' + targetSail.name, {
                exit: 1,
                errorId: 'SERVE-APPSAIL-2',
                arg: [
                    (0, ansi_colors_1.underline)(((_d = targetSail.config) === null || _d === void 0 ? void 0 : _d.raw.build_path) || ''),
                    ansi_colors_1.italic.underline(targetSail.config.raw.source_path || ''),
                    ansi_colors_1.italic.underline(((_e = targetSail.config) === null || _e === void 0 ? void 0 : _e.raw.build_path) || ''),
                    ansi_colors_1.italic.underline(target)
                ]
            });
        }
        switch (true) {
            case (0, container_1.isContainer)(): {
                child = yield startAppSailWithContainer(serverDetails);
                break;
            }
            case (_g = (_f = targetSail.config) === null || _f === void 0 ? void 0 : _f.stack) === null || _g === void 0 ? void 0 : _g.startsWith('node'): {
                child = yield startAppSail(targetSail.port.appsail.host, {
                    type: 'nodejs',
                    target,
                    command: targetSail.config.command,
                    memory: targetSail.config.memory || 256,
                    env: targetSail.config.env_variables,
                    name: targetSail.name
                });
                break;
            }
            case (_j = (_h = targetSail.config) === null || _h === void 0 ? void 0 : _h.stack) === null || _j === void 0 ? void 0 : _j.startsWith('python'): {
                child = yield startAppSail(targetSail.port.appsail.host, {
                    type: 'python',
                    target,
                    command: targetSail.config.command,
                    memory: targetSail.config.memory || 256,
                    env: targetSail.config.env_variables,
                    name: targetSail.name
                });
                break;
            }
            case (_l = (_k = targetSail.config) === null || _k === void 0 ? void 0 : _k.stack) === null || _l === void 0 ? void 0 : _l.startsWith('java'): {
                if (targetSail.config.platform === 'war') {
                    const jettyPath = (0, path_1.join)(constants_1.ENVPATH.runtimes.data, 'jetty');
                    if (!(yield fs_1.ASYNC.fileExists((0, path_1.join)(jettyPath, 'start.jar')).catch((err) => (0, index_1.debug)(err)))) {
                        try {
                            yield fs_1.ASYNC.ensureDir(jettyPath);
                            const appxAPI = yield (0, endpoints_1.appSailAPI)();
                            const jettyZip = yield appxAPI.downloadJetty();
                            const zip = new archiver_1.default();
                            zip.load(jettyZip);
                            zip.extract(jettyPath, '/', {
                                isFolder: true
                            });
                            yield zip.finalize();
                        }
                        catch (err) {
                            (0, index_1.debug)('Unable to ensure Jetty Runtime');
                            throw err;
                        }
                    }
                    const jettyCommand = `java -jar "${(0, path_1.join)(jettyPath, 'start.jar')}" -Djetty.deploy.monitoredPath="${(0, project_1.resolveProjectPath)(target)}" -Djetty.http.port=${targetSail.port.appsail.host}`;
                    child = yield startAppSail(targetSail.port.appsail.host, {
                        target,
                        type: 'war',
                        command: jettyCommand,
                        memory: targetSail.config.memory || 256,
                        env: Object.assign({ JETTY_BASE: (0, path_1.join)(jettyPath, 'JETTY_BASE') }, targetSail.config.env_variables),
                        name: targetSail.name
                    });
                    break;
                }
                child = yield startAppSail(targetSail.port.appsail.host, {
                    target,
                    command: targetSail.config.command,
                    type: 'javase',
                    memory: targetSail.config.memory || 256,
                    env: targetSail.config.env_variables,
                    name: targetSail.name
                });
                break;
            }
            default: {
                const appSailRuntime = yield (0, common_1.getRuntimeDetails)();
                throw new error_1.default('Invalid AppSail stack', {
                    exit: 1,
                    errorId: 'SERVE-APPSAIL-4',
                    arg: [
                        (0, ansi_colors_1.bold)(targetSail.name),
                        ansi_colors_1.bold.red(((_m = targetSail.config) === null || _m === void 0 ? void 0 : _m.stack) + ''),
                        appSailRuntime.runtimes.map((val) => (0, ansi_colors_1.bold)(`* ${val}`)).join('\n')
                    ]
                });
            }
        }
    }
    else if (targetSail.runtime === 'custom') {
        child = yield startAppSailWithContainer(serverDetails);
    }
    else {
        throw new error_1.default('Invalid runtime value for AppSail: ' + targetSail.runtime, {
            exit: 2
        });
    }
    const masterServe = yield (0, master_1.default)(targetSail.port.proxy, {
        appSailDetails: serverDetails
    });
    masterServe.on('listening', () => {
        utils_1.serverEvent.emit('start', serverDetails);
    });
    masterServe.on('error', (err) => {
        (0, index_1.debug)('Error with the AppSail proxy server: ', err);
    });
    child.once('exit', (code) => __awaiter(void 0, void 0, void 0, function* () {
        var _o, _p, _q, _r, _s, _t, _u;
        typeof targetSail.port.appsail === 'number' &&
            port_resolver_1.default.freePort(targetSail.port.appsail);
        port_resolver_1.default.freePort(targetSail.port.proxy);
        if (code === 150) {
            targetSail.validity = {
                valid: false,
                reason: 'Unable to start the AppSail'
            };
            throw new error_1.default('Unable to start the AppSail', {
                exit: 1,
                errorId: 'SERVE-APPSAIL-3',
                arg: [
                    (0, ansi_colors_1.bold)(targetSail.name),
                    ((_o = targetSail.config) === null || _o === void 0 ? void 0 : _o.command) || '',
                    (0, ansi_colors_1.underline)(((_p = targetSail.config) === null || _p === void 0 ? void 0 : _p.build_path) || ''),
                    ((_q = targetSail.config) === null || _q === void 0 ? void 0 : _q.stack) || ''
                ]
            });
        }
        if ((_s = (_r = targetSail.config) === null || _r === void 0 ? void 0 : _r.scripts) === null || _s === void 0 ? void 0 : _s.postserve) {
            yield executeHook((_u = (_t = targetSail.config) === null || _t === void 0 ? void 0 : _t.scripts) === null || _u === void 0 ? void 0 : _u.postserve, `AppSail [POSTSERVE] [${targetSail.name}]`, targetSail.source);
        }
    }));
    return child;
});
