// zcatalyst-cli/lib/event_generate/stratus.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getData = exports.sourceId = void 0;
const moment_1 = __importDefault(require("moment"));
const runtime_store_1 = __importDefault(require("../runtime-store"));
const js_1 = require("../util_modules/js");
const project_1 = require("../util_modules/project");
exports.sourceId = 'NULL';
function getData() {
    return __awaiter(this, void 0, void 0, function* () {
        const userDetails = {
            email_id: runtime_store_1.default.get('user.Email'),
            user_type: 'Admin',
            user_id: js_1.JS.randomNumber(15),
            is_confirmed: getRandomBoolean(),
            last_name: js_1.JS.randomString(5),
            first_name: js_1.JS.randomString(10),
            zuid: js_1.JS.randomNumber(10)
        };
        const time = (0, moment_1.default)().format('MMM DD, YYYY hh:mm A');
        return {
            action_time: js_1.JS.randomNumber(15),
            bucket_details: {
                project_details: {
                    id: (0, project_1.getProjectId)(),
                    project_name: (0, project_1.getProjectName)(),
                    project_type: 'Live'
                },
                created_time: time,
                modified_time: time,
                bucket_name: 'sample-bucket',
                modified_by: userDetails,
                created_by: userDetails,
                bucket_meta: {
                    versioning: getRandomBoolean(),
                    encryption: getRandomBoolean(),
                    audit_consent: getRandomBoolean(),
                    caching: {
                        status: getRandomBoolean() ? 'Enabled' : 'Disabled'
                    }
                }
            },
            object_details: [
                {
                    key: 'sample.txt'
                }
            ]
        };
    });
}
exports.getData = getData;
function getRandomBoolean() {
    return js_1.JS.random(0, 1) === 1;
}
