// zcatalyst-cli/lib/event_generate/user.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getData = exports.sourceId = void 0;
const moment_1 = __importDefault(require("moment"));
const runtime_store_1 = __importDefault(require("../runtime-store"));
const js_1 = require("../util_modules/js");
exports.sourceId = 'NULL';
function getData() {
    return __awaiter(this, void 0, void 0, function* () {
        const mockTime = (0, moment_1.default)().format('MMM DD, YYYY hh:mm A');
        return {
            platform_type: 'web',
            user_details: {
                email_id: runtime_store_1.default.get('user.Email'),
                created_time: mockTime,
                is_confirmed: js_1.JS.random(0, 1) === 1 ? true : false,
                last_name: '',
                source: '',
                zuid: js_1.JS.randomNumber(10),
                role_details: {
                    role_name: 'App User',
                    role_id: js_1.JS.randomNumber(10)
                },
                zaaid: js_1.JS.randomNumber(10),
                invited_time: mockTime,
                modified_time: mockTime,
                user_type: 'App User',
                user_id: js_1.JS.randomNumber(10),
                org_id: js_1.JS.randomNumber(10),
                first_name: js_1.JS.randomString(10),
                status: 'ACTIVE'
            },
            org_id: js_1.JS.randomNumber(10),
            zaid: js_1.JS.randomNumber(10),
            redirect_url: ''
        };
    });
}
exports.getData = getData;
