// zcatalyst-cli/lib/deploy/features/functions/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fnLogger = void 0;
const ansi_colors_1 = require("ansi-colors");
const runtime_store_1 = __importDefault(require("../../../runtime-store"));
const util_1 = require("../../util");
const logger_1 = require("../../../util_modules/logger");
const fnLogger = () => {
    const _logTagets = runtime_store_1.default
        .get('context.payload.functions.targets', [])
        .reduce((logTargs, target) => {
        target.name = (0, ansi_colors_1.bold)(target.name);
        switch (target.type) {
            case 'aio': {
                if (logTargs.aio.length === 0) {
                    logTargs.aio.push((0, ansi_colors_1.bold)('==> Advanced I/O'));
                }
                logTargs.aio.push({
                    label: 'DEPLOYMENT SUCCESSFUL',
                    message: target.name,
                    logType: 'SUCCESS'
                });
                logTargs.aio.push({
                    label: 'FUNCTION URL         ',
                    message: (0, ansi_colors_1.underline)(target.url || 'Unknown'),
                    logType: 'MESSAGE'
                });
                logTargs.aio.push('');
                break;
            }
            case 'bio': {
                if (logTargs.bio.length === 0) {
                    logTargs.bio.push((0, ansi_colors_1.bold)('==> Basic I/O'));
                }
                logTargs.bio.push({
                    label: 'DEPLOYMENT SUCCESSFUL',
                    message: target.name,
                    logType: 'SUCCESS'
                });
                logTargs.bio.push({
                    label: 'FUNCTION URL         ',
                    message: (0, ansi_colors_1.underline)(target.url || 'Unknown'),
                    logType: 'MESSAGE'
                });
                logTargs.bio.push('');
                break;
            }
            case 'browserlogic': {
                if (logTargs.browserlogic.length === 0) {
                    logTargs.browserlogic.push((0, ansi_colors_1.bold)('==> BrowserLogic'));
                }
                logTargs.browserlogic.push({
                    label: 'DEPLOYMENT SUCCESSFUL',
                    message: target.name,
                    logType: 'SUCCESS'
                });
                logTargs.browserlogic.push({
                    label: 'FUNCTION URL         ',
                    message: (0, ansi_colors_1.underline)(target.url || 'Unknown'),
                    logType: 'MESSAGE'
                });
                logTargs.browserlogic.push('');
                break;
            }
            case 'cron': {
                if (logTargs.cron.length === 0) {
                    logTargs.cron.push((0, ansi_colors_1.bold)('==> CRON'));
                }
                logTargs.cron.push({
                    label: 'DEPLOYMENT SUCCESSFUL',
                    message: target.name,
                    logType: 'SUCCESS'
                });
                logTargs.cron.push('');
                break;
            }
            case 'event': {
                if (logTargs.event.length === 0) {
                    logTargs.event.push((0, ansi_colors_1.bold)('==> EVENT'));
                }
                logTargs.event.push({
                    label: 'DEPLOYMENT SUCCESSFUL',
                    message: target.name,
                    logType: 'SUCCESS'
                });
                logTargs.event.push('');
                break;
            }
            case 'integ': {
                if (logTargs.integ.length === 0) {
                    logTargs.integ.push((0, ansi_colors_1.bold)('==> INTEGRATION'));
                }
                logTargs.integ.push({
                    label: 'DEPLOYMENT SUCCESSFUL',
                    message: target.name,
                    logType: 'SUCCESS'
                });
                logTargs.integ.push('');
                break;
            }
            case 'job': {
                if (logTargs.job.length === 0) {
                    logTargs.job.push((0, ansi_colors_1.bold)('==> JOB'));
                }
                logTargs.job.push({
                    label: 'DEPLOYMENT SUCCESSFUL',
                    message: target.name,
                    logType: 'SUCCESS'
                });
                logTargs.job.push('');
                break;
            }
        }
        return logTargs;
    }, {
        aio: [],
        bio: [],
        browserlogic: [],
        event: [],
        cron: [],
        job: [],
        integ: []
    });
    Object.entries(_logTagets).forEach(([_type, logs]) => {
        if (logs.length === 0) {
            return;
        }
        (0, logger_1.info)(' '.repeat(2) + logs.shift());
        (0, logger_1.info)();
        (0, util_1.logPaddedLabels)(logs, ' '.repeat(4));
    });
};
exports.fnLogger = fnLogger;
