// @zcatalyst/container-plugin/out/components/lock.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _SimpleLock_locks;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleLock = void 0;
const events_1 = require("events");
const logger_1 = __importDefault(require("../logger"));
/**
 * Offers Simple lock function so that more than one process
 * may not be executed at the same time
 */
class SimpleLock {
    constructor() {
        /** Event emitter to lock events */
        this.lockEvents = new events_1.EventEmitter();
        /** locks entry */
        _SimpleLock_locks.set(this, {});
    }
    /**
     * Get a lock
     * @param id unique identifier of the resource to be locked
     * @returns true if a lock is acquired
     */
    getLock(id) {
        if (id in __classPrivateFieldGet(this, _SimpleLock_locks, "f") && __classPrivateFieldGet(this, _SimpleLock_locks, "f")[id]) {
            return false;
        }
        logger_1.default.debug('Lock provided for: ' + id);
        __classPrivateFieldGet(this, _SimpleLock_locks, "f")[id] = true;
        return true;
    }
    /**
     * Wait for a locked component to be released
     * @param id unique identifier of the locked component
     * @returns Promise that resolves when the lock on the component is successfully released
     */
    waitRelease(id) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!__classPrivateFieldGet(this, _SimpleLock_locks, "f")[id]) {
                return;
            }
            return new Promise((res) => {
                logger_1.default.debug('Awaiting lock release for: ' + id);
                const evFn = (_id) => {
                    if (_id === id) {
                        res();
                        this.lockEvents.removeListener('release', evFn);
                    }
                };
                this.lockEvents.on('release', evFn);
            });
        });
    }
    /**
     * Release the lock on a component
     * @param id unique identifier of the component to be released
     * @returns true
     */
    releaseLock(id) {
        if (id in __classPrivateFieldGet(this, _SimpleLock_locks, "f") && __classPrivateFieldGet(this, _SimpleLock_locks, "f")[id]) {
            delete __classPrivateFieldGet(this, _SimpleLock_locks, "f")[id];
        }
        logger_1.default.debug('Lock released for: ' + id);
        this.lockEvents.emit('release', id);
        return true;
    }
}
exports.SimpleLock = SimpleLock;
_SimpleLock_locks = new WeakMap();
