// @zcatalyst/cli-utils/out/env.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.isVsCode = exports.isWindows = exports.isMac = exports.isCI = void 0;
exports.isPrimaryShell = isPrimaryShell;
exports.envOverride = envOverride;
exports.getEnvVariable = getEnvVariable;
exports.isCI = process &&
    !!(process.env.CI || // Travis CI, CircleCI, Cirrus CI, Gitlab CI, Appveyor, CodeShip, dsari
        process.env.CONTINUOUS_INTEGRATION || // Travis CI, Cirrus CI
        process.env.BUILD_NUMBER || // Jenkins, TeamCity
        process.env.RUN_ID || // TaskCluster, dsari
        exports.name ||
        false);
exports.isMac = process && process.platform === 'darwin';
exports.isWindows = process &&
    (process.platform === 'win32' ||
        (process.env.OSTYPE !== undefined && /^(msys|cygwin)$/.test(process.env.OSTYPE)));
exports.isVsCode = process.env.ZCATALYST_VSCODE === 'true';
function isPrimaryShell() {
    if (!process || process.env.CATALYST_SUB_PROCESS) {
        return false;
    }
    if (!exports.isWindows && process.env.SHLVL) {
        return parseInt(process.env.SHLVL) <= 1;
    }
    return true;
}
function envOverride(envname, value) {
    var _a;
    if (process.env[envname] !== undefined && ((_a = process.env[envname]) === null || _a === void 0 ? void 0 : _a.length)) {
        return process.env[envname];
    }
    return value;
}
function getEnvVariable(variable, fallback) {
    return process.env[variable] || fallback;
}
