// zcatalyst-cli/lib/endpoints/lib/appsail.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = __importDefault(require("../../error"));
const api_1 = __importDefault(require("../../internal/api"));
const constants_1 = require("../../util_modules/constants");
const index_1 = require("../../util_modules/logger/index");
class AppSail {
    constructor(projectId, opts) {
        this.opts = opts;
        this.projectId = projectId;
    }
    deploy(sourceFsStream, { stack, name, contentLength, command, memory, platform, envVariables = {}, catalystAuth, loginRedirect, port }) {
        return __awaiter(this, void 0, void 0, function* () {
            const formData = {
                stack,
                name,
                code: sourceFsStream
            };
            if (typeof memory === 'number') {
                formData.memory = memory;
            }
            if (typeof platform === 'string') {
                formData.platform = platform;
            }
            const configuration = {};
            if (typeof command === 'string') {
                configuration.startup_command = command;
            }
            if (typeof envVariables === 'object' && Object.keys(envVariables).length > 0) {
                configuration.environment = {
                    variables: envVariables
                };
            }
            if (typeof catalystAuth === 'boolean') {
                configuration.catalyst_auth = catalystAuth;
                if (typeof loginRedirect === 'string') {
                    configuration.login_redirect = loginRedirect;
                }
            }
            if (typeof port === 'number') {
                configuration.port = port;
            }
            if (Object.keys(configuration).length > 0) {
                formData.configuration = JSON.stringify(configuration);
            }
            const res = yield new api_1.default(this.opts).put(`/baas/v1/project/${this.projectId}/appsail/upsert`, {
                formData,
                json: false,
                log: {
                    uploadProgress: {
                        title: `AppSail[${name}]`,
                        total: contentLength,
                        progressType: 'upload'
                    },
                    stream: sourceFsStream,
                    awaitingSpinner: `deploying Appsail[${name}]`
                }
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, index_1.debug)('save AppSail response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    getAllAppsails() {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get(`/baas/v1/project/${this.projectId}/appsail`);
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, index_1.debug)('Get all Appsail response from server  : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    updateConfig(appSailId, env) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).post(`/baas/v1/project/${this.projectId}/appsail/${appSailId}/configuration`, {
                body: {
                    environment: {
                        variables: env
                    }
                }
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, index_1.debug)('Update configuration response from server  : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    downloadJetty() {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default({
                origin: constants_1.ORIGIN.zohoCDNUS,
                authNeeded: false,
                json: false,
                isExternal: true
            }).get('/catalyst/appsail/1.0.0/jetty_runtime_dependency.zip', { encoding: null });
            if (res.body) {
                return res.body;
            }
            (0, index_1.debug)('Download JETTY server response from server  : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    getSignedUrl(name) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get(`/baas/v1/project/${this.projectId}/appsail/get-signature?name=${name}`);
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, index_1.debug)('Get upload signature response from server  : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    uploadToBucket({ signedUrl, length, stream, name }) {
        return __awaiter(this, void 0, void 0, function* () {
            const url = new URL(signedUrl);
            const res = yield new api_1.default(Object.assign(Object.assign({}, this.opts), { isExternal: true, origin: url.origin, authNeeded: false, json: false })).put(url.pathname + '?' + url.searchParams, {
                log: {
                    uploadProgress: {
                        title: `AppSail[${name}]`,
                        total: length,
                        progressType: 'upload'
                    },
                    skipReqBody: true,
                    stream,
                    awaitingSpinner: `Finishing Appsail[${name}] Upload`
                },
                body: stream,
                headers: {
                    contentType: 'application/octet-stream'
                }
            });
            if (res.status === 200) {
                return;
            }
            (0, index_1.debug)('Upload to bucket response from server  : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    customAppSailCallback(target, objectId) {
        var _a, _b, _c, _d, _e;
        return __awaiter(this, void 0, void 0, function* () {
            const configuration = {};
            if (typeof ((_a = target.config) === null || _a === void 0 ? void 0 : _a.command) === 'string') {
                configuration.startup_command = target.config.command;
            }
            if (typeof ((_b = target.config) === null || _b === void 0 ? void 0 : _b.catalyst_auth) === 'boolean') {
                configuration.catalyst_auth = target.config.catalyst_auth;
                if (typeof target.config.loginRedirect === 'string') {
                    configuration.login_redirect = target.config.loginRedirect;
                }
            }
            if (typeof ((_d = (_c = target.config) === null || _c === void 0 ? void 0 : _c.port) === null || _d === void 0 ? void 0 : _d.http) === 'number') {
                configuration.port = target.config.port.http;
            }
            const res = yield new api_1.default(this.opts).put(`/baas/v1/project/${this.projectId}/appsail/upsert`, {
                formData: {
                    name: target.name,
                    memory: ((_e = target.config) === null || _e === void 0 ? void 0 : _e.memory) || 2048,
                    platform: 'custom_runtime',
                    configuration: JSON.stringify(configuration),
                    local_object_key: objectId
                }
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, index_1.debug)('Custom AppSail upsert/callback response from server  : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
}
exports.default = AppSail;
