// zcatalyst-cli/lib/endpoints/lib/slate.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = __importDefault(require("../../error"));
const api_1 = __importDefault(require("../../internal/api"));
const logger_1 = require("../../util_modules/logger");
class Slate {
    constructor(projectId, opts) {
        this.opts = opts;
        this.projectId = projectId;
    }
    getFrameworks() {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield new api_1.default(this.opts).get('/slate/v1/frameworks', {
                log: {
                    awaitingSpinner: 'Fetching frameworks from the server ...'
                }
            });
            if (response.body && response.body.data) {
                return response.body.data;
            }
            throw new error_1.default('Error while fetching the frameworks', { exit: 2 });
        });
    }
    getAllApps() {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield new api_1.default(this.opts).get(`/slate/v1/project/${this.projectId}/app`, {
                log: {
                    awaitingSpinner: 'Fetching slate apps'
                }
            });
            if (response.body && response.body.data) {
                return response.body.data;
            }
            throw new error_1.default('Error while fetching the frameworks', { exit: 2 });
        });
    }
    downloadTemplate() {
        return __awaiter(this, void 0, void 0, function* () {
            this.opts.origin = 'https://codeload.github.com';
            this.opts.authNeeded = false;
            const response = yield new api_1.default(Object.assign(Object.assign({}, this.opts), { isExternal: true })).get(`/catalystbyzoho/slate/zip/main`, {
                log: {
                    awaitingSpinner: 'Download template from github ...'
                },
                encoding: null
            });
            if (response.body) {
                return response.body;
            }
            throw new error_1.default('Error while downloading the template', { exit: 2 });
        });
    }
    deploy(sourceStream, contentLength, appName, deployId, deploymentName, appId, config, msg) {
        return __awaiter(this, void 0, void 0, function* () {
            const formData = Object.assign(Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({ file: sourceStream }, (appId && { app_id: appId })), (deployId && { build_id: deployId })), (deploymentName && { deployment_name: deploymentName })), (appId || deployId ? {} : { name: appName })), (config && { configuration: config })), (msg && { message: msg }));
            const res = yield new api_1.default(this.opts).post(`/slate/v1/project/${this.projectId}/app/upload`, {
                formData,
                json: false,
                log: {
                    uploadProgress: {
                        title: `Slate[${appName}]`,
                        total: contentLength,
                        progressType: 'upload'
                    },
                    stream: sourceStream,
                    awaitingSpinner: 'deploying slate'
                }
            });
            if (res.body && res.body.data) {
                return res.body.data;
            }
            (0, logger_1.debug)('deploy slate response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
}
exports.default = Slate;
