// zcatalyst-cli/lib/endpoints/lib/user.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const console_1 = require("console");
const index_js_1 = __importDefault(require("../../error/index.js"));
const api_1 = __importDefault(require("../../internal/api"));
const urls_js_1 = __importDefault(require("../../util_modules/constants/lib/urls.js"));
class Project {
    constructor() {
        this.opts = {
            authNeeded: true
        };
    }
    getUserThumb(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get('/file/download', {
                origin: urls_js_1.default.contacts,
                qs: {
                    ID: id,
                    fs: 'thumb'
                },
                json: false,
                encoding: null
            });
            if (res.body) {
                return res.body;
            }
            (0, console_1.debug)('get thumbnail response from server : ' + res.body);
            throw new index_js_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
}
exports.default = Project;
