// zcatalyst-cli/lib/endpoints/lib/zest.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = __importDefault(require("../../error"));
const api_1 = __importDefault(require("../../internal/api"));
const js_1 = require("../../util_modules/js");
const logger_1 = require("../../util_modules/logger");
const project_1 = require("../../util_modules/project");
class Zest {
    constructor(projectId, opts) {
        this.opts = Object.assign(Object.assign({}, opts), { headers: {
                'CATALYST-ORG': (0, project_1.getEnvId)()
            } });
        this.projectId = projectId;
    }
    getSpecifications(query) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const res = yield new api_1.default(this.opts).get(`/zest/v1/project/${this.projectId}/specifications`, {
                    qs: query,
                    log: {
                        awaitingSpinner: 'Getting Zest Specifications'
                    }
                });
                if (res.body || res.status === 204) {
                    return res.body;
                }
                (0, logger_1.debug)('get specifications response from server : ' + res.body);
                throw new error_1.default('Server Error: Unexpected Response from server.', {
                    exit: 2
                });
            }
            catch (err) {
                if (err.status === 429) {
                    const responseData = err.context;
                    const retryAfter = responseData.response.headers['Retry-After'];
                    setTimeout(() => __awaiter(this, void 0, void 0, function* () { return yield this.getSpecifications(query); }), (retryAfter ? parseInt(retryAfter, 10) : 1) * 1000);
                }
                throw err;
            }
        });
    }
    generateFunction(resources, specEnv, name) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).post(`/zest/v1/project/${this.projectId}/specifications/actions/generate_function`, {
                body: {
                    name,
                    stack: specEnv,
                    resources: resources
                },
                log: {
                    awaitingSpinner: `Scheduling Function Generation ${name}`
                }
            });
            if (res.body && res.body.details && res.body.code === 'SCHEDULED') {
                return res.body.details;
            }
            (0, logger_1.debug)('generate function response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    generateAppsail(requestJson) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).post(`/zest/v1/project/${this.projectId}/specifications/actions/generate_appsail`, {
                body: requestJson,
                log: {
                    awaitingSpinner: `Scheduling Appsail Generation${requestJson.name}`
                }
            });
            if (res.body && res.body.details && res.body.code === 'SCHEDULED') {
                return res.body.details;
            }
            (0, logger_1.debug)('generate appsail response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    getScheduledJob(job_id) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(this.opts).get(`/zest/v1/scheduled_jobs/${job_id}`);
            if (res.body) {
                return res.body;
            }
            (0, logger_1.debug)('get scheduled job response from server : ' + res.body);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
    getFile(file_id, source, name) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield new api_1.default(Object.assign(Object.assign({}, this.opts), { json: false })).get(`/zest/v1/files/${file_id}`, {
                encoding: null,
                log: {
                    progress: {
                        title: `${js_1.JS.capitalize(source)} (${name})`
                    }
                }
            });
            if (res.body) {
                return res.body;
            }
            (0, logger_1.debug)('get file response from server : ' + res);
            throw new error_1.default('Server Error: Unexpected Response from server.', {
                exit: 2
            });
        });
    }
}
exports.default = Zest;
