// zcatalyst-cli/lib/serve/server/lib/python/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const runtime_store_1 = __importDefault(require("../../../../runtime-store"));
const http_functions_1 = require("../../../../shell/dependencies/http-functions");
const userConfig_1 = __importDefault(require("../../../../userConfig"));
const constants_1 = require("../../../../util_modules/constants");
const project_1 = require("../../../../util_modules/project");
const shell_1 = require("../../../../util_modules/shell");
const error_1 = __importDefault(require("../../../../error"));
const index_1 = require("../../../../util_modules/logger/index");
const container_1 = require("../../../../util_modules/container");
const container_plugin_1 = require("@zcatalyst/container-plugin");
exports.default = (details) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    if (!details.target) {
        throw new error_1.default('Invalid Python function target', { exit: 2 });
    }
    const projectRoot = runtime_store_1.default.get('project.root');
    const targetSource = (_a = details.target) === null || _a === void 0 ? void 0 : _a.source;
    const stackVersion = (_b = details.target.stack) === null || _b === void 0 ? void 0 : _b.replace('python_', '');
    const codeLocation = details.target.build;
    const debugConfig = details.debugPort !== -1 ? (_d = (_c = details.target) === null || _c === void 0 ? void 0 : _c.additionalInfo) === null || _d === void 0 ? void 0 : _d.debug : undefined;
    if (!codeLocation) {
        throw new error_1.default('Invalid code location for python function', {
            exit: 2,
            context: details
        });
    }
    if ((0, container_1.isContainer)()) {
        (0, index_1.debug)('using container to serve the function: ' + ((_e = details.target) === null || _e === void 0 ? void 0 : _e.name));
        details.target.env_var = Object.assign({ X_ZOHO_CATALYST_ACCOUNTS_URL: constants_1.ORIGIN.auth, X_ZOHO_CATALYST_CONSOLE_URL: constants_1.ORIGIN.admin, X_ZOHO_CATALYST_RESOURCE_ID: (_f = details.target) === null || _f === void 0 ? void 0 : _f.id, X_ZOHO_STRATUS_RESOURCE_SUFFIX: constants_1.ORIGIN.stratusSuffix, X_ZOHO_CATALYST_CODE_LOCATION: '/catalyst', CATALYST_PROJECT_TIMEZONE: (0, project_1.getProjectTimezone)(Intl.DateTimeFormat().resolvedOptions().timeZone), X_ZOHO_CATALYST_ORG: (0, project_1.getEnvId)(), X_ZOHO_ADMIN_CRED_TOKEN: 'Unknown', CATALYST_FUNCTION_TYPE: constants_1.REMOTE_REF.functions.type[(_g = details.target) === null || _g === void 0 ? void 0 : _g.type] }, (_h = details.target) === null || _h === void 0 ? void 0 : _h.env_var);
        const container = yield Promise.resolve().then(() => __importStar(require('@zcatalyst/container-plugin')));
        return container.start({
            details: details,
            projectId: (0, project_1.getProjectId)(),
            ruok: details.debugPort !== undefined && details.debugPort !== -1
                ? container_plugin_1.RUOK.SKIP
                : container_plugin_1.RUOK.FINITE
        });
    }
    const runtimesDir = constants_1.ENVPATH.runtimes.data;
    const opts = [
        (0, path_1.join)(runtimesDir, constants_1.RUNTIME.language.python.value, `zcatalyst_runtime_${stackVersion === null || stackVersion === void 0 ? void 0 : stackVersion.replace('_', '')}`, 'main.py')
    ];
    if (debugConfig) {
        opts.unshift('-m', debugConfig === null || debugConfig === void 0 ? void 0 : debugConfig.debugger, '--listen', details.debugPort + '');
    }
    opts.unshift('-u');
    return new Promise((resolve, reject) => {
        var _a, _b;
        const child = (0, shell_1.spawn)(userConfig_1.default.get(`python${stackVersion}.bin`), opts, {
            cwd: codeLocation,
            stdio: 'pipe',
            detached: true,
            env: Object.assign({ X_ZOHO_CATALYST_IS_LOCAL: 'true', X_ZOHO_CATALYST_FUNCTION_LOADED: 'true', X_ZOHO_CATALYST_ACCOUNTS_URL: constants_1.ORIGIN.auth, X_ZOHO_CATALYST_CONSOLE_URL: constants_1.ORIGIN.admin, X_ZOHO_CATALYST_RESOURCE_ID: (_a = details.target) === null || _a === void 0 ? void 0 : _a.id, X_ZOHO_STRATUS_RESOURCE_SUFFIX: constants_1.ORIGIN.stratusSuffix, X_ZOHO_CATALYST_CODE_LOCATION: targetSource === null || targetSource === void 0 ? void 0 : targetSource.replace(projectRoot, (0, path_1.join)(projectRoot, constants_1.FOLDERNAME.build)), X_ZOHO_CATALYST_SERVER_LISTEN_PORT: details.httpPort.toString(), CATALYST_PORTAL_DOMAIN: constants_1.ORIGIN.iamPortal, CATALYST_PROJECT_TIMEZONE: (0, project_1.getProjectTimezone)(Intl.DateTimeFormat().resolvedOptions().timeZone) }, details.target.env_var)
        }).RAW();
        child.on('spawn', () => __awaiter(void 0, void 0, void 0, function* () {
            var _c;
            yield (0, http_functions_1.checkIfRuntimeServerRunning)(details.httpPort.toString());
            debugConfig &&
                (0, index_1.info)(`\nPython debugger for function ${(_c = details.target) === null || _c === void 0 ? void 0 : _c.name} listening on port: ${details.debugPort}`);
            resolve(child);
        }));
        const errData = [];
        (_b = child.stderr) === null || _b === void 0 ? void 0 : _b.on('data', (chunk) => {
            errData.push(chunk);
        });
        child.on('error', (err) => {
            (0, index_1.info)(Buffer.concat(errData).toString());
            reject(new error_1.default(`Error while starting python runtime server`, {
                original: err,
                exit: 2
            }));
        });
        child.on('exit', (code, signal) => {
            (0, index_1.debug)(`Python child exit: ${code}, ${signal}`);
        });
    });
});
