// zcatalyst-cli/lib/serve/server/lib/node/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const error_1 = __importDefault(require("../../../../error"));
const runtime_store_1 = __importDefault(require("../../../../runtime-store"));
const constants_1 = require("../../../../util_modules/constants");
const index_1 = require("../../../../util_modules/logger/index");
const project_1 = require("../../../../util_modules/project");
const shell_1 = require("../../../../util_modules/shell");
const container_1 = require("../../../../util_modules/container");
const container_plugin_1 = require("@zcatalyst/container-plugin/");
exports.default = (details) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
    if (!details.target) {
        throw new error_1.default('Invalid Node function target', { exit: 2 });
    }
    const projectRoot = runtime_store_1.default.get('project.root');
    const targetSource = (_a = details.target) === null || _a === void 0 ? void 0 : _a.source;
    const codeLocation = (_b = details.target) === null || _b === void 0 ? void 0 : _b.build;
    if ((0, container_1.isContainer)()) {
        (0, index_1.debug)('using container to serve the function: ' + ((_c = details.target) === null || _c === void 0 ? void 0 : _c.name));
        const container = yield Promise.resolve().then(() => __importStar(require('@zcatalyst/container-plugin')));
        details.target.env_var = Object.assign({ X_ZOHO_CATALYST_ACCOUNTS_URL: constants_1.ORIGIN.auth, X_ZOHO_CATALYST_CONSOLE_URL: constants_1.ORIGIN.admin, X_ZOHO_CATALYST_RESOURCE_ID: (_d = details.target) === null || _d === void 0 ? void 0 : _d.id, X_ZOHO_STRATUS_RESOURCE_SUFFIX: constants_1.ORIGIN.stratusSuffix, X_ZOHO_CATALYST_CODE_LOCATION: '/catalyst', CATALYST_PROJECT_TIMEZONE: (0, project_1.getProjectTimezone)(Intl.DateTimeFormat().resolvedOptions().timeZone), X_ZOHO_CATALYST_ORG: (0, project_1.getEnvId)(), X_ZOHO_ADMIN_CRED_TOKEN: 'Unknown', CATALYST_FUNCTION_TYPE: constants_1.REMOTE_REF.functions.type[(_e = details.target) === null || _e === void 0 ? void 0 : _e.type] }, (_f = details.target) === null || _f === void 0 ? void 0 : _f.env_var);
        return container.start({
            details: details,
            projectId: (0, project_1.getProjectId)(),
            ruok: details.debugPort !== undefined && details.debugPort !== -1
                ? container_plugin_1.RUOK.SKIP
                : container_plugin_1.RUOK.FINITE
        });
    }
    const label = `${details.type}[${(_g = details.target) === null || _g === void 0 ? void 0 : _g.name}]`;
    const _o = details.target, { ['watcher']: _ } = _o, serialisableTarget = __rest(_o, ['watcher']);
    serialisableTarget.index = serialisableTarget.index
        ? serialisableTarget.index.replace(projectRoot, (0, path_1.join)(projectRoot, constants_1.FOLDERNAME.build))
        : undefined;
    const opts = [
        (0, path_1.join)(__dirname, 'server'),
        details.httpPort + '',
        JSON.stringify({
            index: ((_h = details.target) === null || _h === void 0 ? void 0 : _h.index)
                ? details.target.index.replace(projectRoot, (0, path_1.join)(projectRoot, constants_1.FOLDERNAME.build))
                : undefined,
            name: (_j = details.target) === null || _j === void 0 ? void 0 : _j.name,
            type: (_k = details.target) === null || _k === void 0 ? void 0 : _k.type,
            source: (_l = details.target) === null || _l === void 0 ? void 0 : _l.source
        })
    ];
    if (details.debugPort !== -1) {
        (0, index_1.labeled)(label, 'you can attach your debugger at port : ' + details.debugPort).MESSAGE();
        opts.unshift('--inspect=' + details.debugPort);
    }
    return (0, shell_1.spawn)('node', opts, {
        cwd: codeLocation,
        stdio: 'pipe',
        env: Object.assign({ NODE_PATH: targetSource + path_1.sep + 'node_modules', X_ZOHO_CATALYST_IS_LOCAL: 'true', X_ZOHO_CATALYST_FUNCTION_LOADED: 'true', X_ZOHO_CATALYST_ACCOUNTS_URL: constants_1.ORIGIN.auth, X_ZOHO_CATALYST_CONSOLE_URL: constants_1.ORIGIN.admin, X_ZOHO_CATALYST_RESOURCE_ID: (_m = details.target) === null || _m === void 0 ? void 0 : _m.id, X_ZOHO_STRATUS_RESOURCE_SUFFIX: constants_1.ORIGIN.stratusSuffix, CATALYST_PORTAL_DOMAIN: constants_1.ORIGIN.iamPortal, CATALYST_PROJECT_TIMEZONE: (0, project_1.getProjectTimezone)(Intl.DateTimeFormat().resolvedOptions().timeZone), CHROME_EXECUTABLE_PATH: runtime_store_1.default.get('CHROME_EXECUTABLE', process.env.CHROME_EXECUTABLE_PATH), CHROMEDRIVER_EXECUTABLE_PATH: runtime_store_1.default.get('CHROME_DRIVER_EXECUTABLE', process.env.CHROMEDRIVER_EXECUTABLE_PATH) }, details.target.env_var)
    }).RAW();
});
