// zcatalyst-cli/lib/util_modules/server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPortListening = exports.getHostIP = exports.ConnectionDestroyer = void 0;
const http_proxy_1 = __importDefault(require("http-proxy"));
const net_1 = require("net");
const logger_1 = require("./logger");
const error_1 = __importDefault(require("../error"));
class ConnectionDestroyer {
    constructor(server) {
        this.server = server;
        this.connections = new Set();
        if (this.server instanceof http_proxy_1.default) {
            this.server.on('proxyReq', (req) => {
                req.on('close', () => this.connections.delete(req));
                this.connections.add(req);
            });
        }
        else {
            this.server.on('connection', (sock) => {
                sock.on('close', () => this.connections.delete(sock));
                this.connections.add(sock);
            });
        }
    }
    destroy(terminate = true, err) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                this.connections.forEach((sock) => {
                    sock.destroy(err);
                });
                if (terminate) {
                    if (this.server instanceof http_proxy_1.default) {
                        return this.server.close(() => resolve());
                    }
                    return (this.server.listening &&
                        this.server.close((err) => (err instanceof Error ? reject(err) : resolve())));
                }
                resolve();
            });
        });
    }
}
exports.ConnectionDestroyer = ConnectionDestroyer;
function getHostIP(throwErr = false) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const dns = yield Promise.resolve().then(() => __importStar(require('dns/promises')));
            const os = yield Promise.resolve().then(() => __importStar(require('os')));
            const { address } = yield dns.lookup(os.hostname(), 4);
            return address;
        }
        catch (er) {
            if (throwErr)
                throw er;
            (0, logger_1.debug)('Error getting the host ip: ', er);
        }
        return;
    });
}
exports.getHostIP = getHostIP;
function isPortListening(port, retry = 10, retryInterval = 500) {
    return __awaiter(this, void 0, void 0, function* () {
        (0, logger_1.debug)(`itr: ${retry} :: checking port${port} for listening status`);
        if (typeof port === 'string') {
            port = Number.parseInt(port);
        }
        if (retry <= 0) {
            throw new error_1.default('Port not listening');
        }
        retry--;
        const portPromise = yield new Promise((resolve) => {
            const socket = new net_1.Socket();
            const onError = () => {
                socket.destroy();
                resolve(false);
            };
            socket.setTimeout(500);
            socket.once('error', onError);
            socket.once('timeout', onError);
            socket.connect(port, '127.0.0.1', () => {
                socket.end();
                resolve(true);
            });
        });
        if (portPromise) {
            (0, logger_1.debug)(`itr: ${retry} :: port(${port}) is accepting incoming connections`);
            return;
        }
        (0, logger_1.debug)(`itr : ${retry} :: waiting for ${retryInterval}ms before checking for the port(${port}) status again`);
        yield new Promise((resolve) => setTimeout(resolve, retryInterval));
        return isPortListening(port, retry, retryInterval);
    });
}
exports.isPortListening = isPortListening;
