// zcatalyst-cli/lib/util_modules/config/lib/appsail.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.unlinkAppSail = exports.getAllTargetDetails = exports.getTargetDetails = exports.raw = exports.CONTAINER_IMAGE_PROTOCOLS = void 0;
const path_1 = require("path");
const error_1 = __importDefault(require("../../../error"));
const runtime_store_1 = __importDefault(require("../../../runtime-store"));
const constants_1 = require("../../constants");
const fs_1 = require("../../fs");
const project_1 = require("../../project");
exports.CONTAINER_IMAGE_PROTOCOLS = {
    docker: 'docker://',
    dockerArchive: 'docker-archive://'
};
function validateConfig(source, configJson) {
    if (!configJson) {
        return {
            valid: false,
            reason: 'Config file is empty'
        };
    }
    const sourcePath = (0, path_1.isAbsolute)(source) ? source : (0, path_1.join)((0, project_1.getProjectRoot)(), source);
    configJson.raw.source_path = sourcePath;
    let buildPath;
    if ('build_path' in configJson)
        buildPath = configJson.build_path;
    if ('buildPath' in configJson)
        buildPath = configJson.buildPath;
    if (!buildPath) {
        return {
            valid: false,
            reason: `Invalid Build Path: "${buildPath}"`
        };
    }
    const absolutePath = (0, path_1.isAbsolute)(buildPath) ? buildPath : (0, path_1.resolve)(sourcePath, buildPath);
    configJson.build_path = absolutePath;
    configJson.raw.build_path = buildPath;
    if (configJson.catalyst_auth === true) {
        if (configJson.login_redirect === undefined) {
            return {
                valid: false,
                reason: 'Login Redirect value incorrect! Please check the entered value.'
            };
        }
        else if (configJson.login_redirect.length === 0) {
            return {
                valid: false,
                reason: 'Login Redirect value empty! Please provide a login redirect URL.'
            };
        }
        else if (configJson.login_redirect.toLowerCase().startsWith('http') ||
            configJson.login_redirect.includes('://') ||
            !configJson.login_redirect.startsWith('/') ||
            configJson.login_redirect.includes('//')) {
            return {
                valid: false,
                reason: 'Login Redirect value not a relative URL! Please provide a relative URL.'
            };
        }
    }
    return {
        valid: true
    };
}
function raw(throwError = false) {
    const config = runtime_store_1.default.get('config', null);
    if (config === null) {
        if (throwError) {
            throw new error_1.default(constants_1.FILENAME.config + ' file is required', { exit: 2 });
        }
        return;
    }
    const appsailConfig = config.get('appsail');
    const normalizeArchivePath = (path) => {
        const tarPath = path.replace(exports.CONTAINER_IMAGE_PROTOCOLS.dockerArchive, '');
        const normalized = (0, path_1.normalize)((0, path_1.join)((0, project_1.getProjectRoot)(), tarPath));
        return `${exports.CONTAINER_IMAGE_PROTOCOLS.dockerArchive}${normalized}`;
    };
    return appsailConfig === null || appsailConfig === void 0 ? void 0 : appsailConfig.map((_conf) => {
        if (Object.values(exports.CONTAINER_IMAGE_PROTOCOLS)
            .flat()
            .find((proto) => _conf.source.startsWith(proto))) {
            return {
                name: _conf.name,
                source: _conf.source.startsWith(exports.CONTAINER_IMAGE_PROTOCOLS.dockerArchive)
                    ? normalizeArchivePath(_conf.source)
                    : _conf.source,
                runtime: 'custom',
                validity: {
                    valid: true
                },
                config: {
                    command: _conf.command,
                    memory: _conf.memory || 256,
                    port: _conf.port,
                    catalyst_auth: true,
                    login_redirect: '/',
                    build_path: '',
                    env_variables: {},
                    raw: {},
                    stack: ''
                }
            };
        }
        _conf.source = (0, path_1.isAbsolute)(_conf.source) ? _conf.source : (0, path_1.normalize)(_conf.source);
        return {
            name: _conf.name,
            source: _conf.source,
            runtime: 'catalyst',
            validity: {
                valid: true
            }
        };
    });
}
exports.raw = raw;
function getTargetDetails(name) {
    return __awaiter(this, void 0, void 0, function* () {
        const rawTargets = raw();
        const target = rawTargets === null || rawTargets === void 0 ? void 0 : rawTargets.find((tar) => tar.name === name);
        if (!target) {
            return;
        }
        const appConfigPth = (0, project_1.resolveProjectPath)(target.source, constants_1.FILENAME.app_config);
        const isPath = yield fs_1.ASYNC.fileExists(appConfigPth);
        if (!isPath) {
            target.validity = { valid: false, reason: 'Config file not present' };
            return target;
        }
        const appConfigJson = yield fs_1.ASYNC.readJSONFile(appConfigPth);
        if (!appConfigJson) {
            target.validity = { valid: false, reason: 'Config file is empty' };
            return target;
        }
        return Object.assign(Object.assign(Object.assign({}, target), appConfigJson), { validity: { valid: true } });
    });
}
exports.getTargetDetails = getTargetDetails;
function getAllTargetDetails(throwErr = true) {
    return __awaiter(this, void 0, void 0, function* () {
        const rawTargets = raw();
        if (!rawTargets) {
            if (throwErr) {
                throw new error_1.default('No AppSail targets found');
            }
            return;
        }
        return Promise.all(rawTargets.reduce((resArr, target) => {
            if (target.runtime === 'custom') {
                resArr.push(Promise.resolve(target));
                return resArr;
            }
            const catalystConfigPth = (0, project_1.resolveProjectPath)(target.source, constants_1.FILENAME.app_config);
            const isPath = fs_1.SYNC.fileExists(catalystConfigPth);
            if (!isPath) {
                target.validity = {
                    valid: false,
                    reason: 'Config file not present'
                };
                resArr.push(Promise.resolve(target));
                return resArr;
            }
            resArr.push(new Promise((res) => __awaiter(this, void 0, void 0, function* () {
                try {
                    const configJson = yield fs_1.ASYNC.readJSONFile(catalystConfigPth);
                    configJson && (configJson.raw = {});
                    target.validity = validateConfig(target.source, configJson);
                    target.config = configJson;
                    return res(target);
                }
                catch (er) {
                    const configErr = error_1.default.getErrorInstance(er, { skipHelp: true });
                    target.validity = {
                        valid: false,
                        reason: configErr.message
                    };
                    res(target);
                }
            })));
            return resArr;
        }, []));
    });
}
exports.getAllTargetDetails = getAllTargetDetails;
function unlinkAppSail(name) {
    return __awaiter(this, void 0, void 0, function* () {
        const _raw = raw();
        const idx = _raw === null || _raw === void 0 ? void 0 : _raw.findIndex((val) => val.name === name);
        if (!idx) {
            throw new error_1.default(`The provided AppSail name ${name} is not linked to the current project directory`, { exit: 1 });
        }
        const config = runtime_store_1.default.get('config', null);
        if (!config) {
            throw new error_1.default('Unable to retrieve config', { exit: 2 });
        }
        const deleted = _raw === null || _raw === void 0 ? void 0 : _raw.splice(idx, 1);
        config.set('appsail', _raw);
        yield config.save();
        return !!deleted;
    });
}
exports.unlinkAppSail = unlinkAppSail;
