// zcatalyst-cli/lib/util_modules/constants/lib/placeholders.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Object.freeze({
    client: {
        port: '{{_PORT_}}',
        package: {
            name: '{{_NAME_}}'
        }
    },
    functions: {
        node_package: {
            name: '{{_NAME_}}',
            main: '{{_MAIN_}}',
            author: '{{_AUTHOR_}}',
            dependencies: '{{_DEPENDENCIES_}}'
        },
        java_class: '{{_CLASS_}}',
        java_name: '{{_NAME_}}',
        python_package: {
            name: '{{_NAME_}}',
            main: '{{_MAIN_}}',
            sdkVersion: '{{_VERSION_}}'
        }
    },
    catalyst_config: {
        type: '{{_TYPE_}}',
        stack: '{{_STACK_}}',
        integ_config: '{{__INTEG_CONFIG__}}',
        integ_handlers: '{{__INTEG_HANDLERS__}}'
    },
    event_data: {
        project_id: '{{__PROJECT_ID__}}',
        project_name: '{{__PROJECT_NAME__}}',
        eventbus_id: '{{__EVENTBUS_ID__}}',
        eventbus_name: '{{__EVENTBUS_NAME__}}',
        action: '{{__ACTION__}}',
        source: '{{__SOURCE__}}',
        source_id: '{{__SOURCE_ID__}}',
        event_time: '{{__EVENT_TIME__}}',
        data: '{{__DATA__}}',
        rule_id: '{{__RULE_ID__}}'
    },
    signal_event_data: {
        version: '{{__VERSION__}}',
        org_id: '{{__ORG_ID__}}',
        project_id: '{{__PROJECT_ID__}}',
        project_name: '{{__PROJECT_NAME__}}',
        environment: '{{__ENVIRONMENT__}}',
        uuid: '{{__UUID__}}',
        source: '{{__SOURCE__}}',
        signal_id: '{{__SIGNAL_ID__}}',
        signal_api_name: '{{__SIGNAL_API_NAME__}}',
        data: '{{__DATA__}}',
        time_in_ms: '{{__TIME_IN_MS__}}',
        rule_id: '{{__RULE_ID__}}',
        target_id: '{{__TARGET_ID__}}',
        attempt: '{{__ATTEMPT__}}'
    },
    init_js: {
        project_id: '{{__PROJECT_ID__}}',
        zaid: '{{__ZAID__}}',
        auth_domain: '{{__AUTH_DOMAIN__}}',
        org_id: '{{__ORG_ID__}}',
        stratus_domain: '{{_STRATUS_DOMAIN__}}'
    }
});
