// zcatalyst-cli/lib/util_modules/constants/lib/urls.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const env_js_1 = require("../../env.js");
const dc_js_1 = require("../../dc.js");
const dc_type_js_1 = __importDefault(require("./dc-type.js"));
class URL {
    static get zohoCDNUS() {
        return (0, env_js_1.envOverride)('ZOHO_CDN', URL._zohoCDNBase);
    }
    static get zohoCDN() {
        return (0, env_js_1.envOverride)('ZOHO_CDN', mapDataCenterUrl(URL._zohoCDNBase));
    }
    static get auth() {
        return (0, env_js_1.envOverride)('CATALYST_AUTH_URL', mapDataCenterUrl(URL._authBase));
    }
    static get iamPortal() {
        return (0, env_js_1.envOverride)('CATALYST_PORTAL_DOMAIN', mapDataCenterUrl(URL._portalBase));
    }
    static get admin() {
        return (0, env_js_1.envOverride)('CATALYST_ADMIN_URL', mapDataCenterUrl(URL._adminBase));
    }
    static get app() {
        return (0, env_js_1.envOverride)('CATALYST_APP_URL', mapDataCenterUrl(URL._appBase));
    }
    static get appSailDomain() {
        return (0, env_js_1.envOverride)('CATALYST_APPSAIL_URL', mapDataCenterUrl(URL._appSailDomainBase));
    }
    static get catalystStatic() {
        return (0, env_js_1.envOverride)('ZOHO_STATIC', mapDataCenterUrl(URL._zohoStaticBase));
    }
    static get console() {
        return (0, env_js_1.envOverride)('CATALYST_CONSOLE_URL', mapDataCenterUrl(URL._consoleBase));
    }
    static get contacts() {
        const dc = (0, dc_js_1.getActiveDCType)();
        if (dc === undefined) {
            return URL._contacts;
        }
        if (dc.value === dc_type_js_1.default.ca.value) {
            return URL._contacts.replace('.zoho.com', '.zohocloud.ca');
        }
        return URL._contacts.replace('.com', dc.ext);
    }
    static get stratusSuffix() {
        return (0, env_js_1.envOverride)('CATALYST_STRATUS_RESOURCE_SUFFIX', mapDataCenterUrl(URL._stratusSuffixBase));
    }
}
exports.default = URL;
URL._authBase = 'https://accounts.zoho.com';
URL._portalBase = 'https://accounts.zohoportal.com';
URL._adminBase = 'https://api.catalyst.zoho.com';
URL._appBase = 'https://catalystserverless.com';
URL._appSailDomainBase = 'https://catalystappsail.com';
URL._zohoStaticBase = 'https://www.zoho.com/catalyst';
URL._consoleBase = 'https://console.catalyst.zoho.com';
URL._stratusSuffixBase = '.zohostratus.com';
URL._zohoCDNBase = 'https://static.zohocdn.com';
URL._contacts = 'https://profile.zoho.com';
URL.external = Object.freeze({
    gitHubAPI: 'https://api.github.com',
    gitHubWeb: 'https://github.com'
});
function mapDataCenterUrl(url) {
    const dc = (0, dc_js_1.getActiveDCType)();
    if (dc === undefined) {
        return url;
    }
    if (dc.value === dc_type_js_1.default.ca.value && url.includes('.zoho.com')) {
        return url.replace('.zoho.com', '.zohocloud.ca');
    }
    return url.replace(dc_type_js_1.default.us.ext, dc.ext);
}
