// zcatalyst-cli/lib/util_modules/logger/index.js
'use strict';
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.labeled = exports.local = exports.error = exports.warning = exports.message = exports.success = exports.debug = exports.info = exports.log = void 0;
const ansi_colors_1 = __importDefault(require("ansi-colors"));
const utils_1 = require("./utils");
const winston_1 = __importDefault(require("./winston"));
const util_1 = require("util");
const env_1 = require("../env");
const os_1 = require("os");
const logStream = env_1.isVsCode && utils_1.LogStreamFactory.getStream();
function _log(level, message, ignoreHttpLogger = false) {
    if (logStream && logStream.writable && level !== 'debug') {
        logStream.write(message.endsWith(os_1.EOL) ? message : message + os_1.EOL);
    }
    winston_1.default.log({ level, message, ignoreHttpLogger });
}
function log(level, message, ignoreHttpLogger) {
    _log(level, message, ignoreHttpLogger);
}
exports.log = log;
function info(msg = '', ...params) {
    _log('info', (0, util_1.format)(msg, ...params));
}
exports.info = info;
function debug(msg = '', ...params) {
    _log('debug', (0, util_1.format)(msg, ...params));
}
exports.debug = debug;
function success(msg, level = 'info', label = '', padLeft = '') {
    const postChar = label === '' ? '' : label + ': ';
    _log(level, padLeft + ansi_colors_1.default.green(utils_1.CHAR.success + ' ' + postChar) + msg);
}
exports.success = success;
function message(msg, level = 'info', label = '', padLeft = '') {
    const postChar = label === '' ? '' : label + ': ';
    _log(level, padLeft + ansi_colors_1.default.cyan(utils_1.CHAR.info + ' ' + postChar) + msg);
}
exports.message = message;
function warning(msg, level = 'warn', label = '', padLeft = '') {
    const postChar = label === '' ? '' : label + ': ';
    _log(level, padLeft + ansi_colors_1.default.yellow(utils_1.CHAR.warning + ' ' + postChar) + msg);
}
exports.warning = warning;
function error(msg, level = 'error', label = '', padLeft = '') {
    const postChar = label === '' ? '' : label + ': ';
    _log(level, padLeft + ansi_colors_1.default.red(utils_1.CHAR.error + ' ' + postChar) + msg);
}
exports.error = error;
exports.local = {
    info: (msg = '') => log('info', msg, true),
    warn: (msg = '') => log('warn', msg, true),
    error: (msg = '') => log('error', msg, true)
};
function labeled(label, line, padLeft) {
    return {
        SUCCESS: () => success(line, 'info', label, padLeft),
        WARN: () => warning(line, 'warn', label, padLeft),
        MESSAGE: () => message(line, 'info', label, padLeft),
        ERROR: () => error(line, 'error', label, padLeft)
    };
}
exports.labeled = labeled;
__exportStar(require("./utils"), exports);
