// zcatalyst-cli/lib/fn-utils/lib/java/index.js
'use strict';
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTemplatePath = exports.validate = exports.extractAllData = exports.projectFileValidation = exports.ensureJavaInvoker = exports.isValidClassName = exports.classPathSep = exports.FN_TYPE = void 0;
const ansi_colors_1 = require("ansi-colors");
const path_1 = require("path");
const xml2js_1 = __importDefault(require("xml2js"));
const index_js_1 = __importDefault(require("../../../error/index.js"));
const index_js_2 = require("../../../util_modules/constants/index.js");
const env_js_1 = require("../../../util_modules/env.js");
const index_js_3 = require("../../../util_modules/fs/index.js");
const js_js_1 = require("../../../util_modules/js.js");
const option_js_1 = require("../../../util_modules/option.js");
const shell_js_1 = require("../../../util_modules/shell.js");
const index_1 = require("../../../util_modules/logger/index");
const ensure_java_userconfig_js_1 = require("./ensure-java-userconfig.js");
const keywords_js_1 = require("./keywords.js");
const compile_js_1 = require("./compile.js");
const classpath_js_1 = require("./classpath.js");
var index_js_4 = require("../../../util_modules/constants/index.js");
Object.defineProperty(exports, "FN_TYPE", { enumerable: true, get: function () { return index_js_4.FN_TYPE; } });
exports.classPathSep = env_js_1.isWindows ? ';' : ':';
function isValidClassName(name) {
    if (name === '') {
        return false;
    }
    if (js_js_1.JS.upperFirst(name) !== name) {
        return false;
    }
    if ((0, keywords_js_1.containsKeyWord)(name)) {
        return false;
    }
    return true;
}
exports.isValidClassName = isValidClassName;
function ensureJavaInvoker(pth, invoker, target) {
    var _a, _b, _c;
    const spawnCommand = (0, ensure_java_userconfig_js_1.getJavaSpawnCommand)('javac', (_a = target.additionalInfo) === null || _a === void 0 ? void 0 : _a.binPath);
    const classPth = pth + '.class';
    const targetVersion = (0, ensure_java_userconfig_js_1.getTargetVersion)(target.stack);
    if (index_js_3.SYNC.fileExists(classPth)) {
        return true;
    }
    index_js_3.SYNC.ensureDir((0, path_1.dirname)(classPth));
    const child = (0, shell_js_1.spawn)(spawnCommand, [
        '-target',
        targetVersion,
        '-source',
        targetVersion,
        '-cp',
        (0, path_1.join)('lib', '*') + module.exports.classPathSep + '.',
        '-g',
        '-d',
        (0, path_1.dirname)(classPth),
        (0, path_1.basename)(invoker)
    ], { cwd: (0, path_1.dirname)(invoker), detached: env_js_1.isWindows, stdio: 'pipe' }).SYNC();
    if (child.status === 1) {
        const errorResponse = [];
        const err = child.stderr.toString();
        compile_js_1.warningMessages.forEach((element) => {
            if (err.includes(element)) {
                errorResponse.push(err.replace(element, ''));
            }
        });
        if ((_c = (_b = target.additionalInfo) === null || _b === void 0 ? void 0 : _b.compilerOut) === null || _c === void 0 ? void 0 : _c.error) {
            target.additionalInfo.compilerOut.error = errorResponse;
        }
    }
    index_js_3.SYNC.copyDir((0, path_1.join)((0, path_1.dirname)(invoker), 'lib'), (0, path_1.join)((0, path_1.dirname)(classPth), 'lib'));
    return true;
}
exports.ensureJavaInvoker = ensureJavaInvoker;
function projectFileValidation(pth, target) {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function* () {
        const fileExists = yield index_js_3.ASYNC.fileExists(pth);
        if (!fileExists) {
            const templatePath = getTemplatePath(target.type, (_b = (_a = target.integ_config) === null || _a === void 0 ? void 0 : _a.at(0)) === null || _b === void 0 ? void 0 : _b.service);
            yield index_js_3.ASYNC.copyFile((0, path_1.join)(templatePath, index_js_2.FILENAME.functions.java_project), (0, path_1.join)(target.source, index_js_2.FILENAME.functions.java_project));
            yield index_js_3.ASYNC.findAndReplace(pth)(index_js_2.PLACEHOLDER.functions.java_name, target.name);
        }
        const content = yield index_js_3.ASYNC.readFile(pth);
        if (content === undefined) {
            throw new index_js_1.default('Content of ' + pth + ' is undefined', { exit: 2 });
        }
        const result = yield xml2js_1.default.parseStringPromise(content);
        const name = js_js_1.JS.get(result, 'projectDescription.name')[0];
        if (name !== target.name) {
            throw new index_js_1.default('project name mismatch in .project file for function ' + target.name, {
                exit: 1,
                errorId: 'JAVA-3',
                arg: [(0, ansi_colors_1.bold)(name), (0, ansi_colors_1.bold)((0, path_1.basename)(pth)), (0, ansi_colors_1.bold)(target.name)]
            });
        }
    });
}
exports.projectFileValidation = projectFileValidation;
function extractAllData(child, name) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((res, rej) => {
            var _a, _b, _c;
            let message = '';
            (_a = child[name]) === null || _a === void 0 ? void 0 : _a.on('data', (chunk) => {
                message += chunk;
            });
            (_b = child[name]) === null || _b === void 0 ? void 0 : _b.on('end', () => {
                res(message);
            });
            (_c = child[name]) === null || _c === void 0 ? void 0 : _c.on('error', (err) => {
                rej(err);
            });
        });
    });
}
exports.extractAllData = extractAllData;
function validate(targets) {
    return __awaiter(this, void 0, void 0, function* () {
        return _validate(targets);
    });
}
exports.validate = validate;
function _validate(targets, ensured = {}, idx = 0) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        if (targets.length < idx + 1) {
            return;
        }
        const currentTarget = targets[idx];
        try {
            const targetSource = currentTarget.source;
            const classpath = (0, path_1.join)(targetSource, index_js_2.FILENAME.functions.java_classpath);
            const projectPath = (0, path_1.join)(targetSource, index_js_2.FILENAME.functions.java_project);
            const libFolder = (0, path_1.join)(targetSource, 'lib');
            const entry = (0, path_1.join)(targetSource, ((_a = currentTarget.index) === null || _a === void 0 ? void 0 : _a.replace(/\./g, path_1.sep)) + '.java');
            const entryFileExists = yield index_js_3.ASYNC.fileExists(entry);
            const classPathExists = yield index_js_3.ASYNC.fileExists(classpath);
            if (!entryFileExists) {
                throw new index_js_1.default((0, path_1.basename)(entry) + ' file is missing', {
                    exit: 1,
                    errorId: 'JAVA-6',
                    arg: [
                        (0, ansi_colors_1.bold)((0, path_1.basename)(entry)),
                        (0, ansi_colors_1.bold)(index_js_2.FILENAME.catalyst_config),
                        (0, ansi_colors_1.bold)(currentTarget.name)
                    ]
                });
            }
            if (!classPathExists) {
                const templatePath = getTemplatePath(currentTarget.type);
                yield index_js_3.ASYNC.copyFile((0, path_1.join)(templatePath, index_js_2.FILENAME.functions.java_classpath), (0, path_1.join)(targetSource, index_js_2.FILENAME.functions.java_classpath));
                yield (0, classpath_js_1.rewriteClasspath)(classpath, libFolder);
            }
            yield (0, classpath_js_1.normaliseClasspath)(classpath, libFolder);
            yield projectFileValidation(projectPath, currentTarget);
            if (!(currentTarget.stack in ensured)) {
                try {
                    const binPath = yield (0, ensure_java_userconfig_js_1.ensureJava)(currentTarget.stack);
                    if (typeof binPath === 'string') {
                        ensured[currentTarget.stack] = binPath;
                    }
                    else {
                        currentTarget.valid = false;
                        currentTarget.failure_reason =
                            'Unable to find the compatible java version. See the logs for more details';
                        ensured[currentTarget.stack] = false;
                    }
                }
                catch (er) {
                    (0, index_1.debug)('Error ensuring java: ', er);
                    currentTarget.valid = false;
                    currentTarget.failure_reason = er.message;
                    ensured[currentTarget.stack] = false;
                    return _validate(targets, ensured, ++idx);
                }
            }
            const binPath = ensured[currentTarget.stack];
            if (typeof binPath === 'boolean') {
                currentTarget.valid = false;
                return _validate(targets, ensured, ++idx);
            }
            currentTarget.additionalInfo = {
                binPath: binPath,
                compilerOut: {
                    error: [],
                    warning: []
                }
            };
            yield (0, compile_js_1.compileTarget)(currentTarget).catch((err) => {
                throw new index_js_1.default('there was a compilation error!', {
                    exit: 1,
                    errorId: 'JAVA-7',
                    original: err,
                    skipHelp: !!(0, option_js_1.getOptionValue)('watch', false)
                });
            });
            return _validate(targets, ensured, ++idx);
        }
        catch (err) {
            currentTarget.valid = false;
            currentTarget.failure_reason = err.message;
            return _validate(targets, ensured, ++idx);
        }
    });
}
function getTemplatePath(fnType, add) {
    switch (fnType) {
        case index_js_2.FN_TYPE.browserLogic: {
            if (Object.keys(index_js_2.TEMPLATE.functions.java.browserlogic).includes('Selenium')) {
                return index_js_2.TEMPLATE.functions.java.browserlogic.Selenium;
            }
            throw new index_js_1.default('Invalid browserlogic type', { exit: 2 });
        }
        case index_js_2.FN_TYPE.integration: {
            if (!add) {
                throw new index_js_1.default('Invalid additional parameter', { exit: 2 });
            }
            const temp = index_js_2.TEMPLATE.functions.java[fnType];
            return temp[add];
        }
        default: {
            return index_js_2.TEMPLATE.functions.java[fnType];
        }
    }
}
exports.getTemplatePath = getTemplatePath;
__exportStar(require("./classpath"), exports);
__exportStar(require("./compile"), exports);
__exportStar(require("./ensure-java-userconfig.js"), exports);
__exportStar(require("./keywords"), exports);
