// @zcatalyst/container-plugin/out/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerEvents = exports.packImage = exports.ruok = exports.RUOK = exports.ping = exports.start = void 0;
// TODO: look into configuring the memory and cpu time
// TODO: look into appsail custom runtime flow and implement the necessary changes
var start_1 = require("./start");
Object.defineProperty(exports, "start", { enumerable: true, get: function () { return start_1.start; } });
__exportStar(require("./error"), exports);
var system_1 = require("./endpoints/system");
Object.defineProperty(exports, "ping", { enumerable: true, get: function () { return system_1.ping; } });
var container_1 = require("./endpoints/container");
Object.defineProperty(exports, "RUOK", { enumerable: true, get: function () { return container_1.RUOK; } });
Object.defineProperty(exports, "ruok", { enumerable: true, get: function () { return container_1.ruok; } });
var image_1 = require("./endpoints/image");
Object.defineProperty(exports, "packImage", { enumerable: true, get: function () { return image_1.packImage; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "ContainerEvents", { enumerable: true, get: function () { return utils_1.ContainerEvents; } });
