// @zcatalyst/container-plugin/out/prepare.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const components_1 = require("./components");
const utils_1 = require("./utils");
const error_1 = __importDefault(require("./error"));
const image_1 = require("./endpoints/image");
/**
 * Prepares the contains to serve the catalyst compute components
 * @param details details of the catalyst targets
 * @param projectId Id of the current Catalyst Project
 * @returns Promise that resolves to the Serve Container options
 */
exports.default = (details, projectId) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
    const uid = Math.random().toString(36).substring(2, 6);
    switch (details.type) {
        case 'functions':
        case 'server': {
            const _details = details;
            const _target = _details.target;
            const stack = _target.stack;
            const envVars = [
                'X_ZOHO_CATALYST_IS_LOCAL=true',
                'X_ZOHO_CATALYST_FUNCTION_LOADED=true',
                'X_ZOHO_CATALYST_LISTEN_PORT=9000',
                'X_ZOHO_SPARKLET_SERVER_LISTEN_PORT=9000',
                // 'X_ZOHO_SPARKLET_STATUS_FD=1',
                'X_ZOHO_SPARKLET_LOG_FD=1',
                'IS_LOCAL=true',
                'X_ZC_INSTANCE_ID=local',
                'X_ZOHO_PROJECT_SECRET_KEY=Local',
                'X_ZOHO_CATALYST_ENVIRONMENT=Development',
                'X_ZOHO_JOBMETA_JOBID=000',
                ...Object.entries(_target.env_var || {}).map(([env_key, env_val]) => `${env_key}=${env_val}`)
            ];
            if (_target.type === 'browserlogic' && stack.startsWith('node')) {
                envVars.push('NODE_PATH=/catalyst/node_modules');
            }
            let ensureOptions = {};
            if (_target.type === 'browserlogic') {
                ensureOptions = {
                    rootFs: 'browser_logic',
                    lang: stack === 'java8'
                        ? 'java8_browser_logic'
                        : stack === 'node16'
                            ? 'node16_browser_logic'
                            : stack,
                    runtime: stack
                };
            }
            else {
                ensureOptions = {
                    rootFs: stack,
                    lang: stack,
                    runtime: stack
                };
            }
            const ensure = (0, components_1.ensureComponents)(ensureOptions);
            const componentPath = yield ensure;
            return {
                name: `function_${projectId}_${_target.name}_${uid}`,
                image: componentPath.rootFsImage,
                envVars,
                command: (0, utils_1.getFnCommand)({
                    stack: _target.stack,
                    fnType: _target.type,
                    debug: details.debugPort && details.debugPort !== -1 ? true : false,
                    maxMem: 256 + '',
                    minMem: 256 + ''
                }),
                volumes: {
                    codeLocation: (yield (0, utils_1.isDir)(_target.build))
                        ? _target.build
                        : (0, path_1.dirname)(_target.build),
                    lang: componentPath.lang,
                    runtime: componentPath.runtime
                },
                httpPort: {
                    host: _details.httpPort,
                    container: 9000
                },
                debugPort: {
                    host: _details.debugPort,
                    container: 9100
                }
            };
        }
        case 'appsail': {
            const _details = details;
            const _target = _details.target;
            const envVars = [
                'X_ZOHO_CATALYST_IS_LOCAL=true',
                'IS_LOCAL=true',
                'X_ZOHO_PROJECT_SECRET_KEY=Local',
                'X_ZOHO_CATALYST_ENVIRONMENT=Development',
                'JETTY_BASE=/var/runtime/',
                ...Object.entries(((_a = _target.config) === null || _a === void 0 ? void 0 : _a.env_variables) || {}).map(([env_key, env_val]) => `${env_key}=${env_val}`)
            ];
            const command = [];
            const containerHttpPort = !(((_c = (_b = _target.port) === null || _b === void 0 ? void 0 : _b.appsail) === null || _c === void 0 ? void 0 : _c.container) && ((_e = (_d = _target.port) === null || _d === void 0 ? void 0 : _d.appsail) === null || _e === void 0 ? void 0 : _e.container) !== -1)
                ? 9000
                : _target.port.appsail.container;
            envVars.push(`X_ZOHO_CATALYST_LISTEN_PORT=${containerHttpPort}`, `X_ZOHO_SPARKLET_SERVER_LISTEN_PORT=${containerHttpPort}`);
            const hostDebugPort = !(((_g = (_f = _target.port) === null || _f === void 0 ? void 0 : _f.debug) === null || _g === void 0 ? void 0 : _g.host) && _target.port.debug.host !== -1)
                ? undefined
                : _target.port.debug.host;
            hostDebugPort &&
                envVars.push(`X_ZOHO_CATALYST_DEBUG_PORT=${_target.port.debug.container || 9100}`);
            if (_target.runtime === 'custom') {
                const image = _target.source.replace('docker://', '');
                const localImage = yield (0, image_1.findImageInLocal)(image);
                if (!localImage) {
                    throw new error_1.default(`The image ${_target.source} cannot be found in the local image registry`);
                }
                ((_h = _target.config) === null || _h === void 0 ? void 0 : _h.command) && command.push(..._target.config.command.split(' '));
                return {
                    name: `appsail_${projectId}_${_target.name}_${uid}`,
                    image,
                    envVars,
                    command,
                    httpPort: _target.port.appsail,
                    debugPort: hostDebugPort
                        ? {
                            host: hostDebugPort,
                            container: _target.port.debug.container
                        }
                        : undefined
                };
            }
            const stack = (_j = _target.config) === null || _j === void 0 ? void 0 : _j.stack;
            if (!stack) {
                throw new error_1.default('Unable to get the stack for the AppSail: ' + _target.name, { context: details });
            }
            const buildPath = ((_k = _target.config) === null || _k === void 0 ? void 0 : _k.buildPath) || ((_l = _target.config) === null || _l === void 0 ? void 0 : _l.build_path);
            if (!buildPath) {
                throw new error_1.default('Invalid build path: ' + buildPath, {
                    context: details
                });
            }
            if (((_m = _target.config) === null || _m === void 0 ? void 0 : _m.platform) === 'war') {
                command.push('/var/lang/bin/java', '-jar', '/var/runtime/start.jar');
            }
            else {
                command.push('sh', '-c', ((_o = _target.config) === null || _o === void 0 ? void 0 : _o.command) || 'echo Invalid command');
            }
            const componentPath = yield (0, components_1.ensureComponents)({
                rootFs: stack,
                lang: stack,
                runtime: ((_p = _target === null || _target === void 0 ? void 0 : _target.config) === null || _p === void 0 ? void 0 : _p.platform) === 'war' ? 'jetty' : undefined
            });
            return {
                name: `appsail_${projectId}_${_target.name}_${uid}`,
                image: componentPath.rootFsImage,
                envVars,
                command,
                volumes: {
                    codeLocation: (yield (0, utils_1.isDir)(buildPath)) ? buildPath : (0, path_1.dirname)(buildPath),
                    lang: componentPath.lang,
                    runtime: componentPath.runtime
                },
                cwd: ((_q = _target === null || _target === void 0 ? void 0 : _target.config) === null || _q === void 0 ? void 0 : _q.platform) === 'war' ? '/var/runtime/JETTY_BASE' : '/catalyst',
                httpPort: _target.port.appsail,
                debugPort: hostDebugPort
                    ? {
                        host: hostDebugPort,
                        container: _target.port.debug.container
                    }
                    : undefined
            };
        }
        default: {
            throw new error_1.default('Unsupported serve component: ' + details.type, {
                context: details
            });
        }
    }
});
